go/kbfs/libfs/leveldb_storage.go

```diff
@@ -0,0 +1,681 @@
+
+// levelDBStorage is a billy-filesystem-backed storage.
+type levelDBStorage struct {
+	fs       billy.Filesystem
+	readOnly bool
+
+	mu      sync.Mutex
+	flock   billy.File
+	slock   *levelDBStorageLock
+	logw    billy.File
+	logSize int64
+	buf     []byte
+	// Opened file counter; if open < 0 means closed.
+	open int
+	day  int
+}
+
+var _ storage.Storage = (*levelDBStorage)(nil)
+
+// OpenLevelDBStorage returns a new billy-filesystem-backed storage
+// implementation of the levelDB storage interface. This also acquires
+// a file lock, so any subsequent attempt to open the same path will
+// fail.
+func OpenLevelDBStorage(bfs billy.Filesystem, readOnly bool) (
+	s storage.Storage, err error) {
+	flock, err := bfs.OpenFile("LOCK", os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			flock.Close()
+		}
+	}()
+	err = flock.Lock()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		logw    billy.File
+		logSize int64
+	)
+	if !readOnly {
+		logw, err = bfs.OpenFile("LOG", os.O_WRONLY|os.O_CREATE, 0644)
+		if err != nil {
+			return nil, err
+		}
+		logSize, err = logw.Seek(0, os.SEEK_END)
+		if err != nil {
+			logw.Close()
+			return nil, err
+		}
+	}
+
+	fs := &levelDBStorage{
+		fs:       bfs,
+		readOnly: readOnly,
+		flock:    flock,
+		logw:     logw,
+		logSize:  logSize,
+	}
+	runtime.SetFinalizer(fs, (*levelDBStorage).Close)
+	return fs, nil
+}
+
+func (fs *levelDBStorage) writeFileSynced(
+	filename string, data []byte, perm os.FileMode) error {
+	f, err := fs.fs.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return err
+	}
+	n, err := f.Write(data)
+	if err == nil && n < len(data) {
+		err = io.ErrShortWrite
+	}
+	if err1 := f.Close(); err == nil {
+		err = err1
+	}
+	if err != nil {
+		return err
+	}
+	return fs.sync()
+}
+
+func (fs *levelDBStorage) Lock() (storage.Locker, error) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	if fs.open < 0 {
+		return nil, storage.ErrClosed
+	}
+	if fs.readOnly {
+		return &levelDBStorageLock{}, nil
+	}
+	if fs.slock != nil {
+		return nil, storage.ErrLocked
+	}
+	fs.slock = &levelDBStorageLock{fs: fs}
+	return fs.slock, nil
+}
+
+func itoa(buf []byte, i int, wid int) []byte {
+	u := uint(i)
+	if u == 0 && wid <= 1 {
+		return append(buf, '0')
+	}
+
+	// Assemble decimal in reverse order.
+	var b [32]byte
+	bp := len(b)
+	for ; u > 0 || wid > 0; u /= 10 {
+		bp--
+		wid--
+		b[bp] = byte(u%10) + '0'
+	}
+	return append(buf, b[bp:]...)
+}
+
+func (fs *levelDBStorage) printDay(t time.Time) {
+	if fs.day == t.Day() {
+		return
+	}
+	fs.day = t.Day()
+	_, _ = fs.logw.Write([]byte("=============== " + t.Format("Jan 2, 2006 (MST)") + " ===============\n"))
+}
+
+func (fs *levelDBStorage) doLog(t time.Time, str string) {
+	if fs.logSize > logSizeThreshold {
+		// Rotate log file.
+		fs.logw.Close()
+		fs.logw = nil
+		fs.logSize = 0
+		err := fs.fs.Rename("LOG", "LOG.old")
+		if err != nil {
+			return
+		}
```

Comment:
This error used to be ignored, which would mean that it'd just continue logging. I suppose it's caught be linter. I guess if rename fails and leave the old file behind under name `LOG`, it'd error below when trying to create the file with O_CREATE anyway, but I thought I'd point it out in case I missed something that matters.

Reply:
Yeah, did this because of the linter.  Seemed like a reasonable thing to fix to me, and I don't see how it could cause problems, but if you feel strongly about it, I can fix it back in the future.